### DISCUSSION TITLE
Question about rounded corner columns

### DISCUSSION THREAD

**Discussion by kevincianfarini** [patrykandpatrick/vico#644] [Questions] [answered]
I'm using Vico to render a bar chart with rounded corner bars. I've noticed there's some fragmenting where the rounded corners are drawn. The below screenshots illustrates this for bars of varying widths. 

<img width="244" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/00b1ec78-394b-4d04-b8f8-69e2d95b1cad">
<img width="354" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/1dad089c-0b2f-46ae-9e96-a6aef0922ee5">
<img width="484" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/c66149d8-6816-4262-9473-633634805207">

Is there anything I should be doing to ensure the smoothest possible corners on these bars? Here are the shapes I'm currently using to draw my lines. 

```kt
val positiveRoundedCornerShape = Shapes.roundedCornerShape(topLeft = 3.dp, topRight = 3.dp)
val negativeRoundedCornerShape = Shapes.roundedCornerShape(bottomLeft = 3.dp, bottomRight = 3.dp)
val line = LineComponent(
    shape = if (value.toDouble() < 0.0) negativeRoundedCornerShape else positiveRoundedCornerShape
)
```

**Comment by Gowsky** [comment]
Hello, please provide a MRE. It doesn't happen in the sample app with the same `Shape`.
![image](https://github.com/patrykandpatrick/vico/assets/14221573/77c7244e-6c87-4954-90ff-2ef1e6857518)

**Reply by kevincianfarini** [reply]
I was able to reproduce this with the following change in the sample: 

```patch
# sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart2.kt#L97
- shape = Shapes.roundedCornerShape(allPercent = 40),
+ shape = Shapes.roundedCornerShape(topLeft = 3.dp, topRight = 3.dp),
```

<img width="474" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/f751417a-a1ff-4816-a765-fd3739ccb552">

I'm running this on a Pixel 3a API 29 emulator. 

```
Properties
avd.ini.displayname              Pixel 3a API 29
avd.ini.encoding                 UTF-8
AvdId                            Pixel_3a_API_29
disk.dataPartition.size          6442450944
fastboot.chosenSnapshotFile      
fastboot.forceChosenSnapshotBoot no
fastboot.forceColdBoot           no
fastboot.forceFastBoot           yes
hw.accelerometer                 yes
hw.arc                           false
hw.audioInput                    yes
hw.battery                       yes
hw.camera.back                   virtualscene
hw.camera.front                  emulated
hw.cpu.ncore                     4
hw.device.hash2                  MD5:0e6953ebf01bdc6b33a2f54746629c50
hw.device.manufacturer           Google
hw.device.name                   pixel_3a
hw.dPad                          no
hw.gps                           yes
hw.gpu.enabled                   yes
hw.gpu.mode                      auto
hw.initialOrientation            Portrait
hw.keyboard                      yes
hw.lcd.density                   440
hw.lcd.height                    2220
hw.lcd.width                     1080
hw.mainKeys                      no
hw.ramSize                       1536
hw.sdCard                        yes
hw.sensors.orientation           yes
hw.sensors.proximity             yes
hw.trackBall                     no
image.androidVersion.api         29
image.sysdir.1                   system-images/android-29/google_apis_playstore/arm64-v8a/
PlayStore.enabled                true
runtime.network.latency          none
runtime.network.speed            full
showDeviceFrame                  yes
skin.dynamic                     yes
tag.display                      Google Play
tag.id                           google_apis_playstore
vm.heapSize                      228
```

**Reply by kevincianfarini** [reply]
Perhaps this is an Android API level issue? Here's the sample unchanged run in the above emulator. 
<img width="661" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/df5ac4f3-2731-415a-a72c-724958be6137">

**Reply by Gowsky** [reply]
I took my screenshot on an emulator as well, but I've never seen this issue before neither on real devices, nor emulators. My first guess is that your GPU does not play well enough with your emulator. 

Can you see if you have similar issues in other apps? Can you also check whether you have the same problem on a real device?

**Reply by kevincianfarini** [reply]
Sure. I'm also going to try on an API 34 emulator.

**Reply by kevincianfarini** [reply]
There's no artifacting on an API 34 emulator. 
<img width="608" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/ad64e887-66f5-4758-9122-068e3c6f6b27">

**Reply by kevincianfarini** [reply, marked as answer]
Looks like this is perhaps an anomaly of the Pixel 3a API 29 emulator. My real device, a Pizel 3a XL API 29, shows no signs of artifacting either. 
<img width="500" alt="image" src="https://github.com/patrykandpatrick/vico/assets/13242326/001ada2c-caf0-4b4d-9b0d-768552fe1f65">
